### DISCUSSION TITLE
How to scope preflight.css? Needed for WordPress admin / Gutenberg 🤔

### DISCUSSION THREAD

**Discussion by davidwebca** [tailwindlabs/tailwindcss#8240] [Help]
Hi!

Since nesting tailwind @base is not possible anymore with JIT, is there a way to scope only preflight styles? I know how to scope with the important feature, but it only does it for Tailwind's utilities. I'm asking specifically for preflight. Tailwind's utilities need preflight to show correctly some values (border resets, box model resets, etc.) otherwise they get overridden by WordPress's admin styles. I guess that could be useful for other needs, but that's the one I'm facing right now.

Ideally, I would like to avoid copying preflight.css into my own project files, hence why I'm asking but otherwise it's what I'm going to do for now. This question could also be considered as an idea since it would be nice to have a way to do that (if it doesn't already exist).

**Comment by formicagreen** [comment]
Having the exact same issue at the moment, wanting to use Tailwind in WordPress Gutenberg modules.

Is there any solution to this currently?

**Comment by formicagreen** [comment]
I managed to fix this using [postcss-editor-styles](https://github.com/m-e-h/postcss-editor-styles) in case anyone else is reading this.

**Comment by davidwebca** [comment]
Didn't know about the one @formicagreen mentioned, but I solved it on my end with a similar solution with [postcss-prefix-selector](https://github.com/RadValentin/postcss-prefix-selector/) - contrary to other similar solutions (there are many), this one allows to specify which files you want to target and transform some selectors to avoid doubling selectors. My postcss config for this one is this: 

```js
module.exports = {
  plugins: {
    'postcss-import': {},
    'tailwindcss/nesting': {},
    'tailwindcss': {},
    'postcss-preset-env': {},
    'autoprefixer': {},
    'postcss-prefix-selector': {
      prefix: '.editor-styles-wrapper .is-root-container.block-editor-block-list__layout',
      includeFiles: ['editor.css'],
      transform: function(prefix, selector, prefixedSelector, filepath) {
        if (selector.match(/^(html|body)/)) {
          return selector.replace(/^([^\s]*)/, `$1 ${prefix}`);
        } else if(selector.match(/^\.(wp-admin|editor-styles-wrapper)/)) {
          return selector;
        } else {
          return prefixedSelector;
        }
      }
    }
  }
}
```

1) If anything in the CSS file targets html or body, we transform it by adding the selector after so ```body #selector``` becomes ```body .editor-styles-wrapper #selector```
2) If something's already targeting .wp-admin, we assume it's already scoped properly and cancel the scoping by returning the unprefixed selector
3) Otherwise return the normal scoped selector

But in all honesty, I'd still like Tailwind to have a native option for something like that 🤔

**Comment by kohheepeace** [comment]
+1 for this feature. My use-case is also for Wordpress Gutenberg. 


I have also found another demands for this feature and solutions.
- https://www.swyx.io/tailwind-unreset
- https://levelup.gitconnected.com/scope-tailwind-css-base-styles-using-custom-preflight-34f71952e070

@adamwathan Is it possible to make the following CSS work?
```css
:not(.editor-styles-wrapper) {
  @tailwind base;
}
```
Currently it shows below errors
```sh
17:15:22 vite.1    | [vite:css] Nested @tailwind rules were detected, but are not supported.
17:15:22 vite.1    | Consider using a prefix to scope Tailwind's classes: https://tailwindcss.com/docs/configuration#prefix
17:15:22 vite.1    | Alternatively, use the important selector strategy: https://tailwindcss.com/docs/configuration#selector-strategy
17:15:22 vite.1    | 14 |  @import url("https://rsms.me/inter/inter.css");
17:15:22 vite.1    | 15 |  :not(.editor-styles-wrapper) {
17:15:22 vite.1    | 16 |    @tailwind base;
17:15:22 vite.1    |    |     ^
17:15:22 vite.1    | 17 |  }
17:15:22 vite.1    | 18 |  
17:15:23 vite.1    | [vite:css] Nested @tailwind rules were detected, but are not supported.
17:15:23 vite.1    | Consider using a prefix to scope Tailwind's classes: https://tailwindcss.com/docs/configuration#prefix
17:15:23 vite.1    | Alternatively, use the important selector strategy: https://tailwindcss.com/docs/configuration#selector-strategy
17:15:23 vite.1    | 14 |  @import url("https://rsms.me/inter/inter.css");
17:15:23 vite.1    | 15 |  :not(.editor-styles-wrapper) {
17:15:23 vite.1    | 16 |    @tailwind base;
17:15:23 vite.1    |    |     ^
17:15:23 vite.1    | 17 |  }
17:15:23 vite.1    | 18 |  
```

**Reply by kohheepeace** [reply]
There is already discussion for this https://github.com/tailwindlabs/tailwindcss/discussions/7050

**Reply by kohheepeace** [reply]
I extracted [preflight.css](https://unpkg.com/tailwindcss@3.2.0/src/css/preflight.css) and disable preflight then did like below

in `some.scss` file
```scss
:not(.editor-styles-wrapper) {
  ...preflight.css here
}
```

This did not work well because nested css rules over evaluated. (for example text-2xl not working well)

```html
<h3 class="text-2xl leading-6 font-semibold text-on-background">
```

<img width="1438" alt="Screen Shot 2022-10-21 at 17 39 27" src="https://user-images.githubusercontent.com/29557494/197152752-a80c452d-600e-411f-b19f-e3c7202b334c.png">
*image shows "nested css rules `:not(.editor-styles-wrapper)` over evaluated" than `text-2xl`

**Reply by samyarkd** [reply]
so basically i should paste all the preflight.css code inside the scope and than it will be enabled only for that scoe
